spec/integration_spec.cr

```diff
@@ -195,31 +195,85 @@ describe LuckyRouter do
   end
 
   describe "route with trailing slash" do
-    router = LuckyRouter::Matcher(Symbol).new
-    router.add("get", "/users/:id", :show)
-
     context "is defined with a trailing slash" do
-      router.add("get", "/users/", :index)
-
       it "should treat it as a index route when called without a trailing slash" do
+        router = LuckyRouter::Matcher(Symbol).new
+        router.add("get", "/users/", :index)
         router.match!("get", "/users").payload.should eq :index
       end
 
       it "should treat it as a index route when called with a trailing slash" do
+        router = LuckyRouter::Matcher(Symbol).new
+        router.add("get", "/users/", :index)
         router.match!("get", "/users/").payload.should eq :index
       end
     end
 
     context "is defined without a trailing slash" do
-      router.add("get", "/users", :index)
-
       it "should treat it as a index route when called without a trailing slash" do
+        router = LuckyRouter::Matcher(Symbol).new
+        router.add("get", "/users", :index)
         router.match!("get", "/users").payload.should eq :index
       end
 
       it "should treat it as a index route when called with a trailing slash" do
+        router = LuckyRouter::Matcher(Symbol).new
+        router.add("get", "/users", :index)
         router.match!("get", "/users/").payload.should eq :index
       end
     end
   end
+
+  describe "duplicate route checking" do
+    it "raises on normal duplicate" do
+      router = LuckyRouter::Matcher(Symbol).new
+      router.add("get", "/posts/something", :post_index)
+
+      expect_raises LuckyRouter::DuplicateRouteError do
+        router.add("get", "/posts/something", :other_post_index)
+      end
+    end
+
+    it "does not raise on duplicate path with different methods" do
+      router = LuckyRouter::Matcher(Symbol).new
+      router.add("get", "/posts/something", :post_index)
+      router.add("post", "/posts/something", :create_something)
+    end
+
+    it "raises even if path variable is different" do
+      router = LuckyRouter::Matcher(Symbol).new
+      router.add("get", "/posts/:id", :post_show)
+
+      expect_raises LuckyRouter::DuplicateRouteError do
+        router.add("get", "/posts/:post_id", :other_post_show)
+      end
+    end
+
+    it "raises on optional paths when pre-existing path does not have optional part" do
+      router = LuckyRouter::Matcher(Symbol).new
+      router.add("get", "/posts", :post_index)
+
+      expect_raises LuckyRouter::DuplicateRouteError do
+        router.add("get", "/posts/?something", :post_something)
+      end
+    end
+
+    it "raises on optional paths when pre-existing path has optional part" do
+      router = LuckyRouter::Matcher(Symbol).new
+      router.add("get", "/posts/something", :post_something)
+
+      expect_raises LuckyRouter::DuplicateRouteError do
+        router.add("get", "/posts/?something", :post_something_maybe)
```

Comment:
Your example would work. It raises because there is one entry that handles a request to `/posts/something` and another that handles a request to `/posts` AND `/posts/something`.  The optional route would also raise if there was another route to just `/posts`